### ISSUE TITLE
Replace val() with getters/setters

### ISSUE THREAD

**Issue by clabe45** [clabe45/vidar#15]
**`util.val`** is used to sample a property, using keyframes, functions, or just returning a single value.

Since the syntax is gross, let's replace it with setters and getters, which use a similar `util.val` function (to prevent repeated code). Besides that, ~and the fact that each value should be cached per frame~ (I will make this a new issue), the details are left open for now.

As a side benefit, properties like `layer.width` and `layer.height` can default to the movie's width/height right in the getter, or perform similar behavior (instead of doing it manually with every query). More importantly, caching eliminates the bug of multiple different values from function properties per frame.

---

- [ ] Refactor [**movie.js**](src/movie.js)
- [ ] Refactor [**layer.js**](src/layer.js)
- [ ] Refactor [**effect.js**](src/effect.js)

**Comment by alejolale** [comment]
Hi ! 
I'm interested in this work ! could you explain more clear this work ?

Thanks for your answer

**Comment by clabe45** [comment]
Absolutely! Because properties can be [keyframes](https://github.com/clabe45/vidar.js/wiki/Keyframes), [functions](https://github.com/clabe45/vidar.js/wiki/Functions), or just simple values; we need a way to get the *current* value of a property for a frame. Right now, the util function `val(property, element, time)` does this, but I think there is a way to make the syntax nicer (and more friendly to usage from the user).

The task is to write a getter and setter for each public property (that's intended to be used externally). I'm thinking something like:

```js
// for layers
set propName(value) {
    this._propName = value;
}

get propName() {
    return val(this._propName, this, this._movie.currentTime - this.startTime);
}
```

Doing this will solve another problem: There are some properties that require custom logic when accessing their values, like layer dimensions default to the movie's dimensions when they are undefined. Right now, this is handled by resizing the canvas to the current frame's values each frame. Then, the canvas's dimensions are read for the rest of the frame. It would be a lot better to put this logic in a getter, like:

```js
// in layer.Base
get width() {
    return val(this._width, this, this._movie.currentTime - this.startTime);
}
```

I just created #20 which ideally should be done first, so you can use `this.currentTime` for movies and layers and `this._target.currentTime` for effects when you pass the time to `val`.

**Comment by clabe45** [comment]
I updated the OP with a checklist. Just pick one, unless you're feeling ambitious

**Comment by rc41186** [comment]
Any progress on this? I can help with this.

**Comment by clabe45** [comment]
@alejolale Are you currently working on this issue?

**Comment by clabe45** [comment]
@rc41186 I forget about this, yes you can work on it if you're still interested

**Comment by clabe45** [comment]
I decided not to do this, because it would be very error-prone to have to implement setters and getters (that call `val`) for every public property in every class. Instead, I implemented *property filters* (0fca188d2e815298bf09b36d85699bc18e1fdc18) that process the results of `val`. You only need to implement property filters when custom logic needs to be applied to the instant value.
